File: typescript/leetcode/0006-zigzag-conversion/solution01.ts
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
export function convert(s: string, numRows: number): string {
  // https://leetcode.com/problems/zigzag-conversion/discuss/3404/Python-O(n)-Solution-in-96ms-(99.43)
  const table = (new Array(numRows)).fill("")

  if (numRows === 1 || s.length < numRows) {
    return s
  }

  let row = 0
  let step = 1
  for (let i = 0; i < s.length; i++) {
    table[row] += s[i]

    if (row === 0) {
      step = 1
    } else if (row === numRows - 1) {
      step = -1
    }

    row += step
  }

  return table.join("")
}

File: python/leetcode/0006-zigzag-conversion/solution01.py
class Solution:
    # https://leetcode.com/problems/zigzag-conversion/discuss/3404/Python-O(n)-Solution-in-96ms-(99.43)
    def convert(self, s: str, numRows: int) -> str:
        table = [""] * numRows

        row = 0
        step = 1

        if numRows == 1 or len(s) < numRows:
            return s

        for x in s:
            table[row] += x

            if row == 0:
                step = 1
            elif row == numRows -1:
                step = -1
            row += step

        return "".join(table)
